qiskit/qpy/common.py

```diff
@@ -0,0 +1,286 @@
+        )
+
+
+class ProgramTypeKey(bytes, Enum):
+    """Typle key enum for program that Qiskit generates."""
+
+    CIRCUIT = b"q"
+
+    @classmethod
+    def assign(cls, obj):
+        """Assign type key to given object.
+
+        Args:
+            obj (any): Arbitrary object to evaluate.
+
+        Returns:
+            ContainerTypeKey: Corresponding key object.
+
+        Raises:
+            QpyError: if object type is not defined in QPY. Likely not supported.
+        """
+        if isinstance(obj, QuantumCircuit):
+            return cls.CIRCUIT
+
+        raise exceptions.QpyError(
+            f"Object type {type(obj)} is not supported in {cls.__name__} namespace."
+        )
+
+
+def read_instruction_param(file_obj):
+    """Read a single data chunk from the file like object.
+
+    Args:
+        file_obj (File): A file like object that contains the QPY binary data.
+
+    Returns:
+        tuple: Tuple of type key binary and the bytes object of the single data.
+    """
+    data = formats.INSTRUCTION_PARAM(
+        *struct.unpack(
+            formats.INSTRUCTION_PARAM_PACK,
+            file_obj.read(formats.INSTRUCTION_PARAM_SIZE),
+        )
+    )
+
+    return data.type, file_obj.read(data.size)
+
+
+def write_instruction_param(file_obj, type_key, data_binary):
+    """Write statically typed binary data to the file like object.
+
+    Args:
+        file_obj (File): A file like object to write data.
+        type_key (Enum): Object type of the data.
+        data_binary (bytes): Binary data to write.
+    """
+    data_header = struct.pack(formats.INSTRUCTION_PARAM_PACK, type_key, len(data_binary))
+    file_obj.write(data_header)
+    file_obj.write(data_binary)
+
+
+def data_to_binary(obj, serializer, **kwargs):
+    """Convert object into binary data with specified serializer.
+
+    Args:
+        obj (any): Object to serialize.
+        serializer (Callable): Serializer callback that can handle input object type.
+        kwargs: Options set to the serializer.
+
+    Returns:
+        bytes: Binary data.
+    """
+    with io.BytesIO() as container:
+        serializer(container, obj, **kwargs)
+        container.seek(0)
+        binary_data = container.read()
+
+    return binary_data
+
+
+def sequence_to_binary(sequence, serializer, **kwargs):
+    """Convert sequence into binary data with specified serializer.
+
+    Args:
+        sequence (Sequence): Object to serialize.
+        serializer (Callable): Serializer callback that can handle input object type.
+        kwargs: Options set to the serializer.
+
+    Returns:
+        bytes: Binary data.
+    """
+    num_elements = len(sequence)
+
+    with io.BytesIO() as container:
+        container.write(struct.pack(formats.SEQUENCE_PACK, num_elements))
+        for datum in sequence:
+            serializer(container, datum, **kwargs)
+        container.seek(0)
+        binary_data = container.read()
+
+    return binary_data
```

Comment:
I'm fine with restricting this to tuple. Here I tried to make this function type agnostic, and offloaded the type handling to its caller. The generic name `Sequence` is to avoid something weird, e.g. deserializing `List` with `Tuple` namedtuple, which should work. If we restrict input type, I can move typecast logic here (but then we may need to write boilerplate code for adding list serializer). AFAIK there is no list type usage in schedule but I'm not sure for future change :)

Reply:
Nah, lets just leave it. It's just a name and there is nothing specific to a tuple for this code. I agree if the need comes up for a list or some other sequence type we can reuse this easily enough. I was more just wondering when I saw this because I didn't remember having generic sequence serialization support